### PR TITLE
[P2P] KISS 4 - Basic nodes bootstrapping [Merge me after #522] - (Issue: #498)

### LINKED ISSUE #498: [P2P] KISS Node bootstrapping
## Objective

Have a mechanism for nodes to initialize their address books when they just joined the network and don't know anything about the world state.

## Origin Document

While developing #416 I encountered the scenario described above and this issue aims at tracking that piece of work (that's already partially done)

## Goals

- Bootstrap nodes list, make it hardcoded and overridable via flags
- Cycle through the bootstrap nodes until you find one that can provide you an addressbook, if it fails, move onto the next one. If bootstrapping is not possible, error out (the node requires operator intervention)
- Depending on the node's state (see issue # TODO), execute the bootstrapping process

## Deliverable

- [ ] Bootstrapping logic with fallback to the next node
- [ ] Integration with the state machine

## Non-goals / Non-deliverables

- ...

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] Scale up the network
- [ ] check the logs for the new node and make sure:
  - [ ] it's transitioning to the bootstrapping state
  - [ ] it's executing bootstrapping
  - [ ] the addressbook is populated with all nodes (including self)
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @deblasis 
**Co-Owners**: @Olshansk

### PR DESCRIPTION
**Pull request by deblasis**
## Description

This PR has been extracted from #491 and is, hopefully, more digestible from a code-review and scope point of view.

This simply adds entrypoints and basic logic for node bootstrapping leveraging the finite state machine

## Issue

Fixes #498 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Ability to fetch an addressbook from a bootstrap node(s) (by convention the first validator in LocalNet)
- It's possible to override bootstrap nodes via CLI flag

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: p2p/event_handler.go [hunk 1]
```diff
@@ -40,6 +41,31 @@ func (m *p2pModule) HandleEvent(event *anypb.Any) error {
...
+		if !ok {
+			return fmt.Errorf("failed to cast event to StateMachineTransitionEvent")
+		}
+
+		if stateMachineTransitionEvent.NewState == string(coreTypes.StateMachineState_P2P_Bootstrapping) {
+			addrBook := m.network.GetAddrBook()
+			if len(addrBook) == 0 {
+				m.logger.Warn().Msg("No peers in addrbook, bootstrapping")
+
+				err := bootstrap(m)
```
Comment (deblasis) [inline p2p/event_handler.go:55]: https://github.com/pokt-network/pocket/pull/491/files#r1108018053

```suggestion
				if err := bootstrap(m); err != nil {
```

File: p2p/event_handler.go [hunk 2]
```diff
@@ -40,6 +41,31 @@ func (m *p2pModule) HandleEvent(event *anypb.Any) error {
...
+				if err != nil {
```
Comment (deblasis) [inline p2p/event_handler.go:56]: https://github.com/pokt-network/pocket/pull/491/files#r1108018053

```suggestion
				if err := bootstrap(m); err != nil {
```

File: p2p/event_handler.go [hunk 3]
```diff
@@ -40,6 +41,30 @@ func (m *p2pModule) HandleEvent(event *anypb.Any) error {
...
+			addrBook := m.network.GetAddrBook()
+			if len(addrBook) == 0 {
+				m.logger.Warn().Msg("No peers in addrbook, bootstrapping")
+
+				if err := bootstrap(m); err != nil {
+					return err
+				}
+			}
+			if !isSelfInAddrBook(m.address, addrBook) {
+				m.logger.Warn().Msg("Self address not found in addresbook, advertising")
```
Comment (deblasis) [inline p2p/event_handler.go:60]: https://github.com/pokt-network/pocket/pull/491/files#r1108018541
> WHy do we only advertise if we're not in the address book?

Good question. It's because the nodes are created with a genesis and if you are node 005+ you have the first 4 but not self. It's something we need to figure out.

File: p2p/event_handler.go [hunk 4]
```diff
@@ -40,6 +41,30 @@ func (m *p2pModule) HandleEvent(event *anypb.Any) error {
 			}
 		}
 
+	case messaging.StateMachineTransitionEventType:
+		stateMachineTransitionEvent, ok := evt.(*messaging.StateMachineTransitionEvent)
```
Comment (deblasis) [inline p2p/event_handler.go:45]: https://github.com/pokt-network/pocket/pull/491/files#r1108017327
> NIT: Create small local helper functions for each event type. Will make it easier as we add more events

I see your point but I'd rather wait and see when that happens. In this particular event, it's likely that we'll have to create logical branches depending on the properties of the events, if it starts becoming hairy, I'd refactor. 

For now I think it's simple enough and gives an idea of what it can do.

WDYT?

Comment (Olshansk) [reply p2p/event_handler.go:45]: Sgtm. My personal preference would be the other way (easier for me to read) but really don't feel strongly on it this early on. Like you said, prob premature

File: runtime/defaults/defaults.go [hunk 5]
```diff
@@ -36,6 +37,15 @@ var (
 	DefaultP2PUseRainTree     = true
 	DefaultP2PConnectionType  = types.ConnectionType_TCPConnection
 	DefaultP2PMaxMempoolCount = uint64(1e5)
+	// DefaultP2PBootstrapNodesCsv is a list of nodes to bootstrap the network with. By convention, for now, the first validator will provide bootstrapping facilities.
+	//
+	// In LocalNet, the developer will have only one of the two stack online, therefore this is also a poor's man way to simulate the scenario in which a boostrap node is offline.
+	//nolint:gocritic // stringConcatSimplify TBD in code review.
```
Comment (Olshansk) [inline runtime/defaults/defaults.go:43]:
```suggestion
	//nolint:gocritic // stringConcatSimplify fmt.Sprintf is clearer for the reader
```

File: p2p/module.go [hunk 6]
```diff
@@ -199,3 +217,99 @@ func (m *p2pModule) handleNetworkMessage(networkMsgData []byte) {
 
 	m.GetBus().PublishEventToBus(&event)
 }
+
+func configureBootstrapNodes(p2pCfg *configs.P2PConfig, m *p2pModule) error {
```
Comment (Olshansk) [inline p2p/module.go:221]: Can we move this into a helper or utils file rather than the core `module.go`?

File: p2p/module.go [hunk 7]
```diff
@@ -199,3 +217,99 @@ func (m *p2pModule) handleNetworkMessage(networkMsgData []byte) {
...
+		}
+	}
+	m.bootstrapNodes = bootStrapNodes
+	return nil
+}
+
+func isValidHostnamePort(str string) bool {
+	pattern := regexp.MustCompile(`^(https?)://([a-zA-Z0-9.-]+):(\d{1,5})$`)
+	matches := pattern.FindStringSubmatch(str)
+	if len(matches) != 4 {
```
Comment (Olshansk) [inline p2p/module.go:245]:
```suggestion
	if matches != nil || len(matches) != 4 {
```

File: p2p/module.go [hunk 8]
```diff
@@ -199,3 +217,99 @@ func (m *p2pModule) handleNetworkMessage(networkMsgData []byte) {
...
+		return false
+	}
+	protocol := matches[1]
```
Comment (Olshansk) [inline p2p/module.go:248]: It's not zero indexisble?

File: p2p/module.go [hunk 9]
```diff
@@ -199,3 +217,99 @@ func (m *p2pModule) handleNetworkMessage(networkMsgData []byte) {
...
+			return true
+		}
+	}
+	return false
+}
+
+func bootstrap(m *p2pModule) error {
+	var (
+		addrBook typesP2P.AddrBook
+	)
```
Comment (Olshansk) [inline p2p/module.go:271]:
```suggestion
	var addrBook typesP2P.AddrBook
```

File: p2p/event_handler.go (hunk 3 above)
Comment (Olshansk) [reply p2p/event_handler.go:60]: 1. We might end up moving the addr book bootstrapping to a config file rather than genesis.
2. Unrelated to (1), there's no harm to advertise even if some part of the network is already aware of your existence

p.logger.Info().Bool("TODO", true).Msg("Advertise self to network")

cc @bryanchriswhite

File: p2p/module.go [hunk 10]
```diff
@@ -199,3 +217,99 @@ func (m *p2pModule) handleNetworkMessage(networkMsgData []byte) {
...
+func isSelfInAddrBook(selfAddr cryptoPocket.Address, addrBook typesP2P.AddrBook) bool {
+	for _, peer := range addrBook {
+		if peer.Address.Equals(selfAddr) {
+			return true
+		}
+	}
+	return false
+}
+
+func bootstrap(m *p2pModule) error {
```
Comment (Olshansk) [inline p2p/module.go:268]: Why not `func (m *p2pModule) bootstrap() error`?

File: p2p/module.go (hunk 6 above)
Comment (Olshansk) [inline p2p/module.go:221]: Why not func (m *p2pModule)...

File: p2p/module.go [hunk 11]
```diff
@@ -199,3 +217,99 @@ func (m *p2pModule) handleNetworkMessage(networkMsgData []byte) {
...
+	var (
+		addrBook typesP2P.AddrBook
+	)
+
+	for _, bootstrapNode := range m.bootstrapNodes {
+		m.logger.Info().Str("endpoint", bootstrapNode).Msg("Attempting to bootstrap from bootstrap node") // TODO: deblasis - fix this. For some reason it's not logging
```
Comment (Olshansk) [inline p2p/module.go:274]: The logger should already be initialized. Try using the global logger to see if it's a general logger thing or a module-specific logger ting

File: p2p/module.go [hunk 12]
```diff
@@ -199,3 +217,99 @@ func (m *p2pModule) handleNetworkMessage(networkMsgData []byte) {
...
+		log.Println("Attempting to bootstrap from bootstrap node: " + bootstrapNode)
+
+		client, err := rpc.NewClientWithResponses(bootstrapNode)
+		if err != nil {
+			continue
+		}
+		healthCheck, err := client.GetV1Health(context.TODO())
+		if err != nil || healthCheck == nil || healthCheck.StatusCode != http.StatusOK {
+			log.Println("Error getting a green health check from bootstrap node: " + bootstrapNode)
```
Comment (Olshansk) [inline p2p/module.go:283]: Let's figure out the log statement issues in this PR

File: p2p/module_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+				p2pCfg: &configs.P2PConfig{
+					BootstrapNodesCsv: "\n\n",
+				},
+				m: &p2pModule{},
+			},
+			wantBootstrapNodes: []string(nil),
+			wantErr:            true,
+		},
+		{
+			name: "invalid hostname:port pattern for bootstrap nodes string should yield an error and return nil",
```
Comment (Olshansk) [inline p2p/module_test.go:89]: Add details "port number too high yields an error and empty list of bootstrap nodes"

You can also easily add a couple tests for the wrong protocol and a negative port number (you have the code in place for it)

File: p2p/module_test.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+				p2pCfg: &configs.P2PConfig{
+					BootstrapNodesCsv: "     ",
+				},
+				m: new(p2pModule),
+			},
+			wantErr:            false,
+			wantBootstrapNodes: defaultBootstrapNodes,
+		},
+		{
+			name: "untrimmed string boostrap nodes should yield no error and return DefaultP2PBootstrapNodes",
```
Comment (Olshansk) [inline p2p/module_test.go:56]: this does not return the default boostrap nodes

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: p2p/module.go (hunk 8 above)
Comment (deblasis) [reply p2p/module.go:248]: the 0th index is the (outermost) group that matches the whole regex

Comment (deblasis) [reply p2p/module.go:248]: More context: 

https://pkg.go.dev/regexp#Regexp.FindStringSubmatch

> If 'Submatch' is present, the return value is a slice identifying the successive submatches of the expression. Submatches are matches of parenthesized subexpressions (also known as capturing groups) within the regular expression, numbered from left to right in order of opening parenthesis. Submatch 0 is the match of the entire expression, submatch 1 is the match of the first parenthesized subexpression, and so on.

File: p2p/module.go (hunk 7 above)
Comment (deblasis) [reply p2p/module.go:245]: I think it's redundant since the `nil` case in included in "not len(4)"

```go
nill := []string(nil)
if len(nill) != 4 {
    fmt.Println("not 4")
}
```
Prints `not 4`

File: p2p/module_test.go (hunk 14 above)
Comment (deblasis) [reply p2p/module_test.go:56]: bad copy-🍝 that helped finding a bug. Now fixed

File: p2p/event_handler.go (hunk 3 above)
Comment (deblasis) [reply p2p/event_handler.go:60]: Agreed

**Review by Olshansk — APPROVED** [review]
A few minor comments (see them as optional), but otherwise, lgtm. A good way to end off a Monday after Valentine's week ;)
